solutions/minwoo/codetree/놀이기구탑승.py

```diff
@@ -0,0 +1,49 @@
+from collections import defaultdict
+n = int(input())
+_map = defaultdict(set)
+ans = 0
+board = [[0]*n for _ in range(n)]
+
+for i in range(n**2):
+    arr = list(map(int,input().split()))
+    for key in arr[1:]:
+        _map[arr[0]].add(key)
+
+def step1():
+    for k,_set in _map.items() : # k 를 앉혀야함.
+        positions = [] # k가 앉을수있는 후보들 
+        for i in range(n):
+            for j in range(n):
+                if board[i][j]:
+                    continue
+                friends = 0
+                empty = 0
+                for dy,dx in ((0,1),(0,-1),(1,0),(-1,0)):
+                    ny = i+dy
+                    nx = j+dx
+                    if 0<=ny<n and 0<=nx<n :
+                        if board[ny][nx] in _set:
+                            friends+=1
+                        if board[ny][nx] == 0 :
+                            empty+=1
+                positions.append([friends,empty,i,j]) # 후보삽입 .
+        positions.sort(key = lambda x : (-x[0],-x[1],x[2],x[3]))
+        friends, empty, i, j= positions[0]
+        board[i][j] = k
+
+def calc_score():
+    global ans
+    for k,_set in _map.items():
+        for i in range(n):
+            for j in range(n):
+                if board[i][j] == k :
+                    friends = 0
+                    for dy,dx in ((0,1),(0,-1),(1,0),(-1,0)):
+                        ny,nx = dy+i,dx+j
+                        if 0<=ny<n and 0<=nx<n:
+                            if board[ny][nx] in _set :
+                                friends+=1
+                    ans += int(10**(friends-1))
+step1()
+calc_score()
+print(ans)
```

Comment:
```suggestion
def calc_score():
    ans = 0
    for i in range(n):
        for j in range(n):
            friends = 0
            for dy,dx in ((0,1),(0,-1),(1,0),(-1,0)):
                ny,nx = dy+i,dx+j
                if not(0<=ny<n and 0<=nx<n): continue
                if board[ny][nx] in _map[board[i][j]] :
                    friends+=1
            ans += int(10**(friends-1))
    return ans
step1()
print(calc_score())
```

1. global 변수를 피하고 계산된 값을 리턴하는 코드로 refactor 해보았습니다.
2. 불필요한 루프 반복을 줄이고, 최대한 indent를 줄이도록 refactor 해보았습니다.

저는 최적 위치 찾을 때 min, max를 이용해 매번 갱신하는 방식을 많이 써왔는데 구현이 조금 번거롭더라고요. 민우님 코드 보면 일단 다 후보로 올린 다음에 한 번의 sort로 깔끔하게 최적값을 구하는 코드를 잘 쓰시는 것 같아요. 많이 배웁니다!! 👍👍

Reply:
피드백 감사합니다! 너무 인덴트가 깊어서 continue로 early 리턴 하는게 좋겠네요 !!

칭찬 감사합니다. 매번 min max 하는게 시간효율이 가장 좋지만 시간복잡도 안에 들면 왠만하면 쉽게 풀려고 하고있어요 ㅎㅎ 그러다보니 sort를 자주활용하게되네요